README.md
# Platform jumping game

Tässä tasohyppelypelissä perusideana on yrittää päästä mahdollisimman korkealle putoamatta, hyppelemällä tasolta toiselle sekä kerätä kolikoita matkalla.

## Dokumentaatio

* [Vaatimusmäärittely](https://github.com/lahlint/ot-harjoitustyo/blob/main/dokumentaatio/vaatimusmaarittely.md)
* [Käyttöohje](https://github.com/lahlint/ot-harjoitustyo/blob/main/dokumentaatio/kayttoohje.md)
* [Työaikakirjanpito](https://github.com/lahlint/ot-harjoitustyo/blob/main/dokumentaatio/tyoaikakirjanpito.md)
* [Arkkitehtuurikuvaus](https://github.com/lahlint/ot-harjoitustyo/blob/main/dokumentaatio/arkkitehtuuri.md)
* [Testausdokumentti](https://github.com/lahlint/ot-harjoitustyo/blob/main/dokumentaatio/testausdokumentti.md)
* [Changelog](https://github.com/lahlint/ot-harjoitustyo/blob/main/dokumentaatio/changelog.md)

## Ohjelman käynnistäminen ja komennot

Voit ladata viimeisimmän releasen lähdekoodin [täältä](https://github.com/lahlint/ot-harjoitustyo/releases/tag/Loppupalautus) valitsemalla Assets-osion alta Source code.

### Käynnistäminen

1. Asenna riippuvuudet komennolla:

```poetry install```

2. Käynnistä sovellus komennolla:

```poetry run invoke start```

### Testaus

Testien suoritus tapahtuu komennolla:

```poetry run invoke test```

### Testikattavuus

Testikattavuusraportin voi luoda htmlcov-hakemistoon komennolla:

```poetry run invoke coverage-report```

### Pylint

Pylint-tarkistukset voi suorittaa komennolla:

```poetry run invoke lint```


dokumentaatio/changelog.md
## Viikko 4

- Lisätty Game, Player ja Platform -luokat(keskeinen sovelluslogiika)
- Käyttäjä näkee pelihahmon ja tasoja joille pelihahmo voi hyppiä
- Pelin perustoiminnallisuudet toteutettu
    - aloitusnäkymässä on ohjeet pelaamiseen ja voi aloittaa uuden pelin
    - pelihahmolla voi hyppiä tasoja pitkin loputtomasti ylöspäin ja tasot siirtyvät sitä mukaan alaspäin
    - ylälaidassä näkyy pisteet ja ohjeet aloitusnäkymään palaamiseen
    - kun putoaa tulee gameover-näkymä jossa näkyy pisteet ja voi aloittaa uuden pelin
    - alustava muuttuva tausta

## Viikko 5

- Lisätty BoostPlatform-luokka ja sen toiminnallisuudet
    - tämän luokan tasoja tulee tavallisten tasojen lisäksi, mutta harvemmin ja niiltä pääsee hyppäämään tavallista korkeammalle
- BoostPlatform-luokan testiluokka ja testit lisätty

## Viikko 6

- Lisätty Coin-luokka ja sen toiminnallisuudet
    - pelissä voi kerätä kolikoita
    - kerättyjen kolikoiden määrän voi nähdä pelin aikana ylälaidassa sekä gameover-näkymässä
- Coin-luokan testiluokka ja testit lisätty

## Viikko 7

- Highscore lisätty
    - parhaan tuloksen voi nähdä startmenu- ja gameover-näkymissä
    - highscore tallentuu tiedostoon ja luetaan sieltä tarvittaessa
- Most coins lisätty
    - suurimman määrän kerättyjä kolikoita voi nähdä nähdä startmenu- ja gameover-näkymissä
    - tallentuu highscoren tapaan tiedostoon


src/tests/coin_test.py
import unittest
from sprites.coin import Coin
from game import Game


class TestCoin(unittest.TestCase):
    def setUp(self):
        self.coin = Coin(200,200)

    def test_coin_moves_down(self):
        self.coin.update(5)
        self.assertNotEqual(self.coin.rect.y, 200)
    
    def test_coin_is_deleted_after_it_leaves_screen(self):
        self.coin.rect.y = 605
        self.coin.update(0)
        self.assertTrue(self.coin, False)


src/tests/player_test.py
import unittest
from sprites.player import Player
from game import Game


class TestPlayer(unittest.TestCase):
    def setUp(self):
        self.player = Player()

    def test_falling_causes_gameover(self):
        self.player.rect.y = 605
        self.player.update()

        self.assertEqual(self.player.gameover, True)

    def test_player_rect_moves_left(self):
        self.player.rect.x = 200
        self.player.move_l = True
        self.player.update()
        self.assertNotEqual(self.player.rect.x, 200)

    def test_player_rect_moves_right(self):
        self.player.rect.x + self.player.image.get_width() == 200
        self.player.move_r = True
        self.player.update()
        self.assertNotEqual(self.player.rect.x, 200)

    def test_player_past_border_is_true_after_passing_border(self):
        self.player.rect.y = 100
        self.player.update()
        self.assertEqual(self.player.past_border, True)


src/tests/platform_test.py
import unittest
from sprites.platform import Platform
from game import Game


class TestPlatform(unittest.TestCase):
    def setUp(self):
        self.platform = Platform(100,100,64)

    def test_platform_moves_down(self):
        self.platform.rect.y = 200
        self.platform.update(5)
        self.assertNotEqual(self.platform.rect.y, 200)

    def test_platform_is_deleted_after_it_leaves_screen(self):
        self.platform.rect.y = 605
        self.platform.update(0)
        self.assertTrue(self.platform, False)

src/game.py
import random
import pygame
from sprites.player import Player
from sprites.platform import Platform
from sprites.coin import Coin
import game_functions


class Game:
    """Class for handling game variables, gameloop, events and drawing screen.

    Attributes:
        display: sets display
        player: creates the player
        platforms: sprite group for platforms
        platform: creates first platform and is used for spacing platforms away from each other
        coin: is used for spacing coins away from each other
        coins: sprite group for coins
        clock: clock
        scroll: determines how much everything needs to move down based on how player advances
        gameover: if true, gameover-view is shown
        start: if true, start menu is shown
        highest: helps with determining score
        score: shows how far up player got to
        coins_collected: coins collected
        boost_timer: helps with spacing out boost platforms
    """
    def __init__(self, start=True):
        """Class constructor that sets game variables and starts game

        Args:
            start: determines whether start menu is shown 
        """

        pygame.init()
        self.display = pygame.display.set_mode((400, 600))
        pygame.display.set_caption("Platform jumping game")
        self.player = Player()
        self.platforms = pygame.sprite.Group()
        self.platform = Platform(163, 560, 40)
        self.platforms.add(self.platform)
        self.coin = Coin(random.randint(0,400-32),random.randint(0,400))
        self.coins = pygame.sprite.Group()
        self.clock = pygame.time.Clock()
        self.scroll = 0
        self.gameover = self.player.gameover
        self.start = start
        self.highest = 0
        self.score = 0
        self.coins_collected = 0
        self.boost_timer = 0

        self.gameloop()

    def new_game(self, start):
        """Starts new game.

        Args:
            start: determines whether start menu is shown 
        """

        Game(start)

    def gameloop(self):
        """Gameloop.
        """

        while True:
            game_functions.check_events(self)
            self.player.update()
            game_functions.check_collisions(self, self.platforms, self.coins, self.player)
            game_functions.scrolling_and_score(self)
            game_functions.check_gameover(self)
            game_functions.draw_screen(self)
            self.clock.tick(60)

if __name__ == "__main__":
    Game()


src/ui/ui.py
from tkinter import Tk, ttk, constants


class UI:
    def __init__(self, root):
        self._root = root

    def start(self):
        heading_label = ttk.Label(master=self._root, text="Kirjautuminen")
        username_label = ttk.Label(master=self._root, text="Käyttäjänimi")
        username_entry = ttk.Entry(master=self._root)
        password_label = ttk.Label(master=self._root, text="Salasana")
        password_entry = ttk.Entry(master=self._root)
        button = ttk.Button(master=self._root, text="Kirjaudu")

        heading_label.grid(row=0, column=0, columnspan=2)
        username_label.grid(row=1, column=0)
        username_entry.grid(row=1, column=1)
        password_label.grid(row=2, column=0)
        password_entry.grid(row=2, column=1)
        button.grid(row=3, column=0, columnspan=2)


window = Tk()
window.title("Tasohyppelypeli")

ui = UI(window)
ui.start()


window.mainloop()


src/sprites/coin.py
import os
import pygame


path = "ot-harjoitustyo/src/sprites/" #pylint: disable=C0103
dirname = os.path.dirname(__file__)


class Coin(pygame.sprite.Sprite):
    """Class for coins, that player can collect.

    Attributes:
        image: coin image
        rect: rectangle, the size of the image, for collisions
        rect.x: x-coordinate of coin's rect
        rect.y: y-coordinate of coin's rect
    """

    def __init__(self, c_x, c_y):
        """Class constructor that creates a new coin.

        Args:
            c_x: x-coordinate of coin
            c_y: y-coordinate of coin
        """

        super().__init__()

        self.image = pygame.image.load(os.path.join(
            dirname, "..", "assets", "coin.png"))
        self.rect = self.image.get_rect()
        self.rect.x = c_x
        self.rect.y = c_y

    def update(self, scroll):
        """Updates coin's position accordingly and deletes coin when needed.

        Args:
            scroll: determines how much coin needs to go down depending on how player advances
        """
        self.rect.y += scroll

        if self.rect.y > 600:
            self.kill()
            

src/sprites/player.py
import os
import pygame


path = "ot-harjoitustyo/src/sprites/" #pylint: disable=C0103
dirname = os.path.dirname(__file__)


class Player(pygame.sprite.Sprite):
    """Class for player/character.

    Attributes:
        image: character image
        rect: rectangle, the size of the image, for collisions
        rect.x: x-coordinate of rect
        rect.y: y-coordinate of rect
        move_r: determines whether player will move right
        move_l: determines whether player will move left
        gameover: if true player has fallen off screen
        past_border: player has reached border where scrolling starts 
        speed: determines speed
    """
    def __init__(self, p_x=170, p_y=370):
        """Class constructor that creates player.

        Args:
            p_x: x-coordinate of rect
            p_y: y-coordinate of rect
        """
        super().__init__()

        self.image = pygame.image.load(
            os.path.join(dirname, "..", "assets", "blob.png"))
        self.rect = self.image.get_rect()
        self.rect.x = p_x
        self.rect.y = p_y
        self.move_r = False
        self.move_l = False
        self.gameover = False
        self.past_border = False
        self.speed = 0

    def update(self):
        """Updates players position and checks for falling off screen.
        """
        if self.move_l and self.rect.x > 0:
            self.rect.x -= 5
        if self.move_r and self.rect.x + self.image.get_width() < 400:
            self.rect.x += 5
        gravity = 1
        if self.rect.y + self.image.get_height() < 600:
            self.speed -= gravity
        else:
            self.gameover = True
            self.speed = 0
        self.rect.y -= self.speed
        if self.rect.y <= 200:
            self.past_border = True


src/sprites/platform.py
import os
import pygame


path = "ot-harjoitustyo/src/sprites/" #pylint: disable=C0103
dirname = os.path.dirname(__file__)


class Platform(pygame.sprite.Sprite):
    """Class for regular platforms, that player can jump on.

    Attributes:
        image: platform image
        rect: rectangle, the size of the image, for collisions
        rect.x: x-coordinate of platform's rect
        rect.y: y-coordinate of platform's rect
        width: width of platform
        boost: helps differentiate between different kinds of platforms in other parts of code
    """
    def __init__(self, p_x, p_y, width):
        """Class constructor that creates new platform.

        Args:
            p_x: x-coordinate of platform
            p_y: y-coordinate of platform
            width: platfroms width
        """
        super().__init__()

        self.image = pygame.image.load(os.path.join(
            dirname, "..", "assets", "platform.png"))
        self.rect = self.image.get_rect()
        self.rect.x = p_x
        self.rect.y = p_y
        self.width = width
        self.boost = False

    def update(self, scroll):
        """Updates platforms position accordingly and deletes platform when needed.

        Args:
            scroll: determines how much platform needs to go down depending on how player advances
        """
        self.rect.y += scroll
        if self.rect.y > 600:
            self.kill()


src/sprites/boost_platform.py
import os
import pygame


path = "ot-harjoitustyo/src/sprites/" #pylint: disable=C0103
dirname = os.path.dirname(__file__)


class BoostPlatform(pygame.sprite.Sprite):
    """Class for platforms that give an extra boost for player.

    Attributes:
        image: platform image
        rect: rectangle, the size of the image, for collisions
        rect.x: x-coordinate of platform's rect
        rect.y: y-coordinate of platform's rect
        width: width of platform
        boost: helps differentiate between different kinds of platforms in other parts of code
    """
    def __init__(self, p_x, p_y, width):
        """Class constructor that creates a new platform

        Args:
            p_x: x-coordinate of platform
            p_y: y-coordinate of platform
            width: platfroms width
        """
        super().__init__()

        self.image = pygame.image.load(os.path.join(
            dirname, "..", "assets", "boost_platform.png"))
        self.rect = self.image.get_rect()
        self.rect.x = p_x
        self.rect.y = p_y
        self.width = width
        self.boost = True

    def update(self, scroll):
        """Updates platforms positions accordingly or deletes platform when needed.

        Args:
            scroll: determines how much platform needs to go down depending on how player advances.
        """
        self.rect.y += scroll
        if self.rect.y > 600:
            self.kill()


src/game_functions.py
import random
import os
import sys
import pygame
from sprites.platform import Platform
from sprites.boost_platform import BoostPlatform
from sprites.coin import Coin


def check_events(game):
    """Checks events/user interaction.

    Args:
        game: game object
    """
    for event in pygame.event.get():
        if event.type == pygame.KEYDOWN:
            if event.key == pygame.K_LEFT:
                game.player.move_l = True
            if event.key == pygame.K_RIGHT:
                game.player.move_r = True
        if event.type == pygame.KEYUP:
            if event.key == pygame.K_LEFT:
                game.player.move_l = False
            if event.key == pygame.K_RIGHT:
                game.player.move_r = False

        if event.type == pygame.KEYDOWN:
            if event.key == pygame.K_RETURN:
                game.start = False
                game.new_game(False)
            if event.key == pygame.K_F2:
                game.start = True
                game.new_game(True)

        if event.type == pygame.QUIT:
            sys.exit()

def scrolling_and_score(game):
    """Handles scrolling and determines score

    Args:
        game: game object
    """
    scroll_border = 200
    if game.player.rect.y <= scroll_border:
        if game.player.speed > 0:
            game.scroll = game.player.speed
            game.player.rect.y = game.player.rect.y + game.scroll
            if game.player.past_border:
                game.score += game.scroll/60
    else:
        game.scroll = 0
        if not game.player.past_border:
            if (600-game.player.rect.y)/60 > game.highest:
                game.highest = (600-game.player.rect.y)/60
                game.score = game.highest

def check_gameover(game):
    """Checks whether player has fallen.

    Args:
        game: game object
    """
    if game.player.gameover:
        game.gameover = True

def draw_screen(game):
    """Handles drawing everything on screen.

    Args:
        game: game object
    """
    if game.gameover:
        draw_gameover_screen(game)
    elif game.start:
        draw_startmenu(game)
    else:
        draw_game(game)

    pygame.display.flip()

def draw_gameover_screen(game):
    """Handles drawing gameover-view.

    Args:
        game: game object
    """
    font1 = pygame.font.SysFont("Arial", 28)
    font2 = pygame.font.SysFont("Arial", 45)
    pygame.draw.rect(game.display, (0, 0, 0), (50, (600-210)/2 -50, 300, 210))
    pygame.draw.rect(game.display, (255, 255, 255), (55, (600-200)/2 -50, 290, 200))
    message = font2.render("GAME OVER", True, (0, 0, 0))
    game.display.blit(message, (200-message.get_width()/2, 170))
    messages = [f"SCORE: {game.score:.0f}", f"COINS: {game.coins_collected}", "ENTER = New game"]
    heights = [230, 270, 310]
    blit_messages_not_centered(game, messages, heights, font1, (0, 0, 0), 65)

    pygame.draw.rect(game.display, (0, 0, 0), (50, 365, 300, 80))
    pygame.draw.rect(game.display, (255, 255, 255),
                        (55, 370, 290, 70))
    high_score = highscore(game)
    message = font1.render("HIGHSCORE: " + str(high_score), True, (255, 0, 0))
    game.display.blit(message, (65, 375))
    mostcoins = most_coins(game)
    message = font1.render("MOST COINS: " + str(mostcoins), True, (255, 0, 0))
    game.display.blit(message, (65, 405))

def draw_startmenu(game):
    """Handles drawing startmenu-view.

    Args:
        game: game object
    """
    font1 = pygame.font.SysFont("Arial", 28)
    font2 = pygame.font.SysFont("Arial", 45)
    game.display.fill((187, 255, 255))
    pygame.draw.rect(game.display, (0, 0, 0), (50 -5, (600-210)/2 + 5, 300 +10, 255))
    pygame.draw.rect(game.display, (255, 255, 255), (55 -5, (600-200)/2 + 5, 290 +10, 245))
    pygame.draw.rect(game.display, (0, 0, 0), (50 -5, 465, 300 +10, 80))
    pygame.draw.rect(game.display, (255, 255, 255), (55 -5, 470, 290 +10, 70))
    messages = ["Platform", "jumping", "game!"]
    heights = [30, 80, 130]
    blit_messages(game, messages, heights, font2)
    messages = ["Jump as high on the", "platforms as you can", "and collect coins!"]
    heights = [210, 250, 290]
    blit_messages(game, messages, heights, font1)
    messages = ["Use the ARROW KEYS", "to move sideways.", "ENTER = start game"]
    heights = [330, 370, 410]
    blit_messages(game, messages, heights, font1)
    messages = ["HIGHSCORE: " + str(highscore(game)), "MOST COINS: " + str(most_coins(game))]
    heights = [475, 505]
    blit_messages_not_centered(game, messages, heights, font1, (255, 0, 0), 65)

def draw_game(game):
    """Handles drawing everything during the game.

    Args:
        game: game object
    """
    colors = [(187, 255, 255), (160, 238, 238), (150, 205, 205),
                (123, 185, 200), (79, 148, 205), (0, 154, 205), (10,115,205), (0,75,170)]
    score_tresholds = [0, 100, 300, 500, 700, 900, 1100, 1400]
    for i, score_treshold in enumerate(score_tresholds):
        if game.score > score_treshold:
            color = colors[i]
    game.display.fill(color)
    font1 = pygame.font.SysFont("Arial", 28)

    generate_platforms(game)
    game.platforms.update(game.scroll)
    game.platforms.draw(game.display)
    generate_coins(game)
    game.coins.update(game.scroll)
    game.coins.draw(game.display)
    game.display.blit(game.player.image,
                        (game.player.rect.x, game.player.rect.y))
    message = font1.render(f"Score: {game.score:.0f}", True, (0, 0, 0))
    game.display.blit(message, (10, 5))
    message = font1.render(f"Coins: {game.coins_collected}", True, (0, 0, 0))
    game.display.blit(message, (10, 30))
    message = font1.render("F2 = Menu", True, (0, 0, 0))
    game.display.blit(message, (400-message.get_width() - 10, 5))

def generate_platforms(game):
    """Generates platforms, regular and boost.

    Args:
        game: game object
    """
    while len(game.platforms) < 10:
        width = 64
        p_x = random.randint(0, 400 - width)
        p_y = game.platform.rect.y - random.randint(90, 120)
        game.platform = Platform(p_x, p_y, width)
        game.platforms.add(game.platform)
        game.boost_timer += 1
    if game.boost_timer == random.randint(15, 20):
        game.boost_timer = 0
        width = 64
        p_x = random.randint(0, 400 - width)
        p_y = game.platform.rect.y - random.randint(90, 120)
        game.platform = BoostPlatform(p_x, p_y, width)
        game.platforms.add(game.platform)

def generate_coins(game):
    """Generates coins.

    Args:
        game: game object
    """
    while len(game.coins) < 10:
        c_x = random.randint(0, 400 - 32)
        c_y = game.coin.rect.y - random.randint(300, 700)
        game.coin = Coin(c_x, c_y)
        game.coins.add(game.coin)

def blit_messages(game, messages, heights, font):
    """Helps with blitting multiple lines of text.

    Args:
        game: game object
        messages: list of strings
        heights: list of heights on which the messages will be placed
        font: font with font size
    """
    for i, string in enumerate(messages):
        message = font.render(string, True, (0, 0, 0))
        game.display.blit(message, (200-message.get_width()/2, heights[i]))

def blit_messages_not_centered(game, messages, heights, font, color, position):
    """Helps with blitting multiple lines of text that is not centered.

    Args:
        game: game object
        messages: list of strings
        heights: list of heights on which the messages will be placed
        font: font with font size
        color: font color
        position: x-coordinate
    """
    for i, string in enumerate(messages):
        message = font.render(string, True, color)
        game.display.blit(message, (position, heights[i]))

def check_collisions(game, platforms, coins, player):
    """Checks players collisions with platforms and coins.

    Args:
        game: game object
        platforms: platform group
        coins: coin group
        player: player object
    """
    for platform in platforms:
        if platform.boost and platform.rect.colliderect(player.rect):
            if player.speed < 0:
                player.speed = 45
        else:
            if platform.rect.colliderect(player.rect) and player.speed < 0:
                player.speed = 25

    for coin in coins:
        if coin.rect.colliderect(player.rect):
            coin.kill()
            game.coins_collected += 1

def highscore(game):
    """Reads and writes file with highscore.

    Args:
        game: game object
    """
    if os.path.exists("highscore.txt"):
        with open("highscore.txt", "r", encoding="utf-8") as file:
            high_score = int(file.read())
    else:
        high_score = 0
    if game.gameover and (game.score > high_score):
        high_score = f"{game.score:.0f}"
        with open("highscore.txt", "w", encoding="utf-8") as file:
            file.write(high_score)
    return high_score

def most_coins(game):
    """Reads and writes file with number of most coins collected.

    Args:
        game: game object
    """
    if os.path.exists("most_coins.txt"):
        with open("most_coins.txt", "r", encoding="utf-8") as file:
            mostcoins = int(file.read())
    else:
        mostcoins = 0
    if game.gameover and (game.coins_collected > mostcoins):
        mostcoins = str(game.coins_collected)
        with open("most_coins.txt", "w", encoding="utf-8") as file:
            file.write(mostcoins)
    return mostcoins
